arch/common/multilevel_irq.c

```diff
@@ -6,35 +6,41 @@
  */
 
 #include <zephyr/device.h>
+#include <zephyr/irq.h>
 #include <zephyr/sw_isr_table.h>
+#include <zephyr/sys/__assert.h>
 #include <zephyr/sys/util.h>
 
+BUILD_ASSERT((CONFIG_NUM_2ND_LEVEL_AGGREGATORS * CONFIG_MAX_IRQ_PER_AGGREGATOR) <=
+		     BIT(CONFIG_2ND_LEVEL_INTERRUPT_BITS),
+	     "L2 bits not enough to cover the number of L2 IRQs");
```

Comment:
I think both of these should be in units of "per aggregator". I.e. that it's incorrect to multiply the LHS side of this relationship with `CONFIG_NUM_2ND_LEVEL_AGGREGATORS`.
```suggestion
BUILD_ASSERT(CONFIG_MAX_IRQ_PER_AGGREGATOR <= BIT(CONFIG_2ND_LEVEL_INTERRUPT_BITS),
	     "L2 bits not enough to cover the number of L2 IRQs");
```

Reply:
That's how the offset is calculated for each aggregator in the multi-level LUT:

https://github.com/zephyrproject-rtos/zephyr/blob/9102821fbe282d1b994ee21c0e3015c885de3d58/arch/common/multilevel_irq.c#L48-L55

`CONFIG_MAX_IRQ_PER_AGGREGATOR` basically means "Max IRQs per PLIC" in the context of RISCV PLIC, to get the total number of IRQs supported in the 2nd level in software, we need to multiply `CONFIG_MAX_IRQ_PER_AGGREGATOR` by the number of 2nd level aggregator (`CONFIG_NUM_2ND_LEVEL_AGGREGATORS`)

If we have 1 PLIC that supports 1024 IRQs, we will have 1024 IRQs in the 2nd level, which need 10 bits in the 2nd level (2<sup>10</sup> = 1024) to encode.

If we have another instance of such PLIC, we will have 2048 IRQs in the 2nd level, requiring 11 bits in the 2nd level to encode (2<sup>11</sup> = 2048).

Reply:
Hmm... It's a bit counter intuitive.

Fine for now I guess but when we start looking at chained interrupts, we should probably scrap this